### ISSUE TITLE
Warn on method names overlapping common traits

### ISSUE THREAD

**Issue by Xion** [rust-lang/rust-clippy#2231]
The standard traits in Rust's `std` -- especially those that are implicitly imported as part of the prelude -- are effectively "reserving" a bunch of common method names. When you see `X::from`, for example, you'd absolutely expect it to be an invocation of `<X as From>::from`.

However, it is possible to implement a standalone `from` method that simply overlaps in name:
```rust
struct X;
impl X {
    pub fn from(i: i32) -> String {  // doesn't even have to return Self
        format!("{}", i)
    }
}
```

and the compiler is perfectly happy to accept that. But this is a highly unidiomatic practice, and one I think Clippy should warn about.

The exact list of traits to be affected by such warning can be argued for, of course. In my view, it should include at least the prelude ones, but possibly also other extremely common traits: `Read`, `AsRef`, `Write`, `Borrow`, etc. Some obvious candidates, like `PartialEq`/`Eq` (with `eq` & `ne`) or `Display`/`Debug` (with `fmt`) are less of an issue, though, since almost no code invokes their methods directly, but in principle they could still be included.

The obvious counterargument to the existence of such a lint would be some standard library types which actually violate it. This includes `Option::as_ref`, `RefCell::borrow`, and a couple others. I'm under the impression, however, that many if not all such cases are widely considered a mistake. Indeed, the clash between `RefCell::borrow` and `Borrow::borrow` is a footgun that crops up rather consistently in the code written by less experienced rustaceans.

When it comes to lint name bikeshedding, I propose something like `common_trait_method_shadowing`; feel free to come up with better ideas.

**Comment by Havvy** [comment]
As a secondary, this should also check method names declared in traits for pretty much the same reason.

**Comment by ada4a** [comment]
I think this was done in #228

**Comment by nk9** [comment]
Yes, this seems to have been implemented as [`should_implement_trait`](https://rust-lang.github.io/rust-clippy/stable/index.html#/should_implement_trait)

@rustbot label S-fixed
